src/home_robot/home_robot/utils/voxel.py

```diff
@@ -0,0 +1,381 @@
+            "mean",
+            "max",
+            "sum",
+        ], f"Unknown feature pool method {feature_pool_method}"
+
+        self.reset()
+
+    def reset(self):
+        """Resets internal tensors"""
+        self._points, self._features, self._weights, self._rgb = None, None, None, None
+        self._mins = self.dim_mins
+        self._maxs = self.dim_maxs
+
+    def add(
+        self,
+        points: Tensor,
+        features: Tensor,
+        rgb: Tensor,
+        weights: Optional[Tensor] = None,
+    ):
+        """Add a feature pointcloud to the voxel grid.
+
+        Args:
+            points (Tensor): N x 3 points to add to the voxel grid
+            features (Tensor): N x D features associated with each point.
+                Reduction method can be set with feature_reduciton_method in init
+            rgb (Tensor): N x 3 colors s associated with each point.
+            weights (Optional[Tensor], optional): Weights for each point.
+                Can be detection confidence, distance to camera, etc.
+                Defaults to None.
+        """
+        if weights is None:
+            weights = torch.ones_like(points[..., 0])
+
+        # Update voxel grid bounds
+        # This isn't strictly necessary since the functions below can infer the bounds
+        # But we might want to do this anyway to enforce that bounds are a multiple of self.voxel_size
+        # And to enforce that the added points are within user-defined boundaries, if those were specified.
+        pos_mins, _ = points.min(dim=0)
+        pos_maxs, _ = points.max(dim=0)
+        if self.dim_mins is not None:
+            assert torch.all(
+                self.dim_mins <= pos_mins
+            ), "Got points outside of user-defined 3D bounds"
+        if self.dim_maxs is not None:
+            assert torch.all(
+                pos_maxs <= self.dim_maxs
+            ), "Got points outside of user-defined 3D bounds"
+
+        if self._mins is None:
+            self._mins, self._maxs = pos_mins, pos_maxs
+            # recompute_voxels = True
+        else:
+            assert (
+                self._maxs is not None
+            ), "How did self._mins get set without self._maxs?"
+            # recompute_voxels = torch.any(pos_mins < self._mins) or torch.any(self._maxs < pos_maxs)
+            self._mins = torch.min(self._mins, pos_mins)
+            self._maxs = torch.max(self._maxs, pos_maxs)
+
+        if self._points is None:
+            assert (
+                self._features is None
+            ), "How did self._points get unset while _features is set?"
+            # assert self._rgbs is None, "How did self._points get unset while _rgbs is set?"
+            assert (
+                self._weights is None
+            ), "How did self._points get unset while _weights is set?"
+            all_points, all_features, all_weights, all_rgb = (
+                points,
+                features,
+                weights,
```

Comment:
What is weights supposed to be?

Reply:
It's supposed to be some point/voxel weights when calculating the mean. Ie when you add one frame to the global voxel grid, if a voxel already has aggregated K points you'd want to weight that more heavily

Reply:
But later on we might want the weights to also incorporate detection confidence, distance to the camera, etc.